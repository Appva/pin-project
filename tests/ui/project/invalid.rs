mod argument {
    use pin_project::{pin_project, project};

    #[pin_project]
    struct A<T>(#[pin] T);

    #[project]
    fn unexpected_stmt1() {
        let mut x = A(0);
        #[project()] //~ ERROR unexpected token
        let A(_) = Pin::new(&mut x).project();
    }

    #[project]
    fn unexpected_stmt2() {
        let mut x = A(0);
        #[project(foo)] //~ ERROR unexpected token
        let A(_) = Pin::new(&mut x).project();
    }

    #[project()] // Ok
    fn unexpected_fn1() {}

    #[project(foo)] //~ ERROR unexpected token
    fn unexpected_fn2() {}
}

mod attribute {
    use pin_project::{pin_project, project};

    #[pin_project]
    struct A<T>(#[pin] T);

    #[project]
    fn duplicate_stmt() {
        let mut x = A(0);
        #[project]
        #[project] //~ ERROR duplicate #[project] attribute
        let A(_) = Pin::new(&mut x).project();
    }

    // FIXME: Using #[project] on a function that doesn't contain #[project] is no-op,
    //        but, ideally, it should be detected.
    #[project]
    #[project] // Ok
    fn duplicate_fn() {}
}

fn main() {}
