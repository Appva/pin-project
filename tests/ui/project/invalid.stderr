error: unexpected token: ()
  --> $DIR/invalid.rs:10:18
   |
10 |         #[project()] //~ ERROR unexpected token
   |                  ^^

error: unexpected token: (foo)
  --> $DIR/invalid.rs:17:18
   |
17 |         #[project(foo)] //~ ERROR unexpected token
   |                  ^^^^^

error: unexpected token: foo
  --> $DIR/invalid.rs:24:15
   |
24 |     #[project(foo)] //~ ERROR unexpected token
   |               ^^^

error: duplicate #[project] attribute
  --> $DIR/invalid.rs:38:9
   |
38 |         #[project] //~ ERROR duplicate #[project] attribute
   |         ^^^^^^^^^^
